src/foam/dao/MDAO.java

```diff
@@ -98,19 +96,25 @@ public FObject put_(X x, FObject obj) {
 
     synchronized ( writeLock_ ) {
       FObject oldValue = find_(x, obj);
-      Object  state    = getState();
 
-      if ( oldValue != null ) {
-        state = index_.remove(state, oldValue);
+      if ( oldValue == null ) {
+        Object state = getState();
+        setState(index_.put(state, obj));
+      } else {
+        update(obj, oldValue);
       }
-
-      setState(index_.put(state, obj));
     }
 
     onPut(obj);
     return obj;
   }
 
+  public void update(FObject obj, FObject oldObj) {
```

Comment:
For symmetry, I would move this code out of this method and just in-line where it is called on line 104. The code for put() isn't a separate method, it's just inlined also. And there's no need to call this from anywhere other than the once place it is called.

Reply:
fixed